hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanP.java

```diff
@@ -0,0 +1,232 @@
+    public boolean complete() {
+        return emitFromTraverser(traverser);
+    }
+
+    static class IMapTraverser implements Traverser<Object[]> {
+        private final List<Expression<?>> projections;
+        private final Expression<Boolean> filter;
+        private final KeyValueIterator iteratorExec;
+
+        private final SimpleExpressionEvalContext ctx;
+        private MapScanRow row;
+
+        IMapTraverser(
+                @Nonnull final NodeEngine nodeEngine,
+                @Nonnull final MapScanMetadata scanMetadata,
+                @Nonnull final List<Integer> partitions,
+                @Nonnull SimpleExpressionEvalContext ctx
+
+        ) {
+            this.ctx = ctx;
+            MapService mapService = nodeEngine.getService(MapService.SERVICE_NAME);
+            MapContainer mapContainer = mapService.getMapServiceContext().getMapContainer(scanMetadata.getMapName());
+            this.projections = scanMetadata.getProjects();
+            this.filter = scanMetadata.getFilter();
+            this.iteratorExec = new MapScanExecIterator(mapContainer, partitions.iterator(), ctx.getSerializationService());
+
+            this.row = MapScanRow.create(
+                    scanMetadata.getKeyDescriptor(),
+                    scanMetadata.getValueDescriptor(),
+                    scanMetadata.getFieldPaths(),
+                    scanMetadata.getFieldTypes(),
+                    mapContainer.getExtractors(),
+                    ctx.getSerializationService()
+            );
+
+        }
+
+        @Override
+        public Object[] next() {
+            while (iteratorExec.tryAdvance()) {
+                Object[] rows = prepareRow(
+                        iteratorExec.getKey(),
+                        iteratorExec.getKeyData(),
+                        iteratorExec.getValue(),
+                        iteratorExec.getValueData()
+                );
+                if (rows != null) {
+                    return rows;
+                }
+            }
+            return null;
+        }
+
+        /**
+         * Prepare the row for the given key and value:
+         * 1) Check filter
+         * 2) Extract projections
+         *
+         * @param rawKey       key as object, might be null
+         * @param rawKeyData   key as data, might be null
+         * @param rawValue     value as object, might be null
+         * @param rawValueData value as data, might be null
+         * @return Row that is ready for processing by parent operators or {@code null} if the row hasn't passed the filter.
+         */
+        protected Object[] prepareRow(Object rawKey, Data rawKeyData, Object rawValue, Data rawValueData) {
+            row.setKeyValue(rawKey, rawKeyData, rawValue, rawValueData);
+
+            // Filter.
+            if (filter != null && TernaryLogic.isNotTrue(filter.evalTop(row, ctx))) {
+                return null;
+            }
+
+            HeapRow heapRow = new HeapRow(projections.size());
```

Comment:
Why not simply create `Object[]` instead of wrapping it in `HeapRow`?

Reply:
Good catch, it makes sence.